pkg/client/apiutil/lazyrestmapper.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiutil
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
+	mapper      meta.RESTMapper
+	client      *discovery.DiscoveryClient
+	knownGroups []string
+}
+
+// NewLazyRESTMapper initializes a LazyRESTMapper
+func NewLazyRESTMapper(c *rest.Config) (meta.RESTMapper, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LazyRESTMapper{
+		mapper: restmapper.NewDiscoveryRESTMapper([]*restmapper.APIGroupResources{}),
+		client: dc,
+	}, nil
+}
+
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string) error {
+	m.knownGroups = append(m.knownGroups, groupName)
+
+	groupResources, err := m.getFilteredAPIGroupResources(m.client)
```

Comment:
Why is this method a member of the `LazyRestMapper` when you're also passing an argument including one of the members from the `LazyRestMapper`, I would suggest changing this up slightly

```
groups := sets.NewString(m.knownGroups.List()...)
groups.Add(groupName)

groupResources, err := getFilteredAPIGroupResources(m.client, groups)
handle(err)

m.knownGroups.Add(groupName)
```

This will make it so we only update the known groups on a successful filter, it also decouples the function from the restmapper struct which makes it less likely to start mutating its state and causing data races.